chapter_linear-networks/linear-regression-concise.md

```diff
@@ -1,424 +1,485 @@
-    print(f'epoch {epoch + 1}, loss {l:f}')
-```
-
-```{.python .input}
-#@tab tensorflow
-num_epochs = 3
-for epoch in range(num_epochs):
-    for X, y in data_iter:
-        with tf.GradientTape() as tape:
-            l = loss(net(X, training=True), y)
-        grads = tape.gradient(l, net.trainable_variables)
-        trainer.apply_gradients(zip(grads, net.trainable_variables))
-    l = loss(net(features), labels)
-    print(f'epoch {epoch + 1}, loss {l:f}')
-```
-
-下面我们[**比较生成数据集的真实参数和通过有限数据训练获得的模型参数**]。
-要访问参数，我们首先从`net`访问所需的层，然后读取该层的权重和偏置。
-正如在从零开始实现中一样，我们估计得到的参数与生成数据的真实参数非常接近。
-
-```{.python .input}
-w = net[0].weight.data()
-print(f'w的估计误差： {true_w - d2l.reshape(w, true_w.shape)}')
-b = net[0].bias.data()
-print(f'b的估计误差： {true_b - b}')
-```
-
-```{.python .input}
-#@tab pytorch
-w = net[0].weight.data
-print('w的估计误差：', true_w - d2l.reshape(w, true_w.shape))
-b = net[0].bias.data
-print('b的估计误差：', true_b - b)
-```
-
-```{.python .input}
-#@tab tensorflow
-w = net.get_weights()[0]
-print('w的估计误差：', true_w - d2l.reshape(w, true_w.shape))
-b = net.get_weights()[1]
-print('b的估计误差：', true_b - b)
-```
-
-## 小结
-
-:begin_tab:`mxnet`
-* 我们可以使用Gluon更简洁地实现模型。
-* 在Gluon中，`data`模块提供了数据处理工具，`nn`模块定义了大量的神经网络层，`loss`模块定义了许多常见的损失函数。
-* MXNet的`initializer`模块提供了各种模型参数初始化方法。
-* 维度和存储可以自动推断，但注意不要在初始化参数之前尝试访问参数。
-:end_tab:
-
-:begin_tab:`pytorch`
-* 我们可以使用PyTorch的高级API更简洁地实现模型。
-* 在PyTorch中，`data`模块提供了数据处理工具，`nn`模块定义了大量的神经网络层和常见损失函数。
-* 我们可以通过`_`结尾的方法将参数替换，从而初始化参数。
-:end_tab:
-
-:begin_tab:`tensorflow`
-* 我们可以使用TensorFlow的高级API更简洁地实现模型。
-* 在TensorFlow中，`data`模块提供了数据处理工具，`keras`模块定义了大量神经网络层和常见损耗函数。
-* TensorFlow的`initializers`模块提供了多种模型参数初始化方法。
-* 维度和存储可以自动推断，但注意不要在初始化参数之前尝试访问参数。
-:end_tab:
-
-## 练习
-
-1. 如果将小批量的总损失替换为小批量损失的平均值，你需要如何更改学习率？
-1. 查看深度学习框架文档，它们提供了哪些损失函数和初始化方法？用Huber损失代替原损失，即
-    $$l(y,y') = \begin{cases}|y-y'| -\frac{\sigma}{2} & \text{ if } |y-y'| > \sigma \\ \frac{1}{2 \sigma} (y-y')^2 & \text{ 其它情况}\end{cases}$$
-1. 你如何访问线性回归的梯度？
-
-:begin_tab:`mxnet`
-[Discussions](https://discuss.d2l.ai/t/1782)
-:end_tab:
-
-:begin_tab:`pytorch`
-[Discussions](https://discuss.d2l.ai/t/1781)
-:end_tab:
-
-:begin_tab:`tensorflow`
-[Discussions](https://discuss.d2l.ai/t/1780)
-:end_tab:
+# 线性回归的简洁实现
+:label:`sec_linear_concise`
+
+在过去的几年里，出于对深度学习强烈的兴趣，
+许多公司、学者和业余爱好者开发了各种成熟的开源框架。
+这些框架可以自动化基于梯度的学习算法中重复性的工作。
+在 :numref:`sec_linear_scratch`中，我们只运用了：
+（1）通过张量来进行数据存储和线性代数；
+（2）通过自动微分来计算梯度。
+实际上，由于数据迭代器、损失函数、优化器和神经网络层很常用，
+现代深度学习库也为我们实现了这些组件。
+
+在本节中，我们将介绍如何(**通过使用深度学习框架来简洁地实现**)
+ :numref:`sec_linear_scratch`中的(**线性回归模型**)。
+
+## 生成数据集
+
+与 :numref:`sec_linear_scratch`中类似，我们首先[**生成数据集**]。
+
+```{.python .input}
+from d2l import mxnet as d2l
+from mxnet import autograd, gluon, np, npx
+npx.set_np()
+```
+
+```{.python .input}
+#@tab pytorch
+from d2l import torch as d2l
+import numpy as np
+import torch
+from torch.utils import data
+```
+
+```{.python .input}
+#@tab tensorflow
+from d2l import tensorflow as d2l
+import numpy as np
+import tensorflow as tf
+```
+

```

Comment:
This part should not be in the diff. We need to add only the paddle code blocks. Not change what was already there. I don't see any change here but it still shows up in the diff for some reason.